packages/vite/src/node/ssr/runtime/esmRunner.ts

```diff
@@ -0,0 +1,121 @@
+import {
+  ssrDynamicImportKey,
+  ssrExportAllKey,
+  ssrImportKey,
+  ssrImportMetaKey,
+  ssrModuleExportsKey,
+} from './constants'
+import type {
+  ResolvedResult,
+  SSRImportMetadata,
+  ViteModuleRunner,
+  ViteRuntimeModuleContext,
+} from './types'
+
+// eslint-disable-next-line @typescript-eslint/no-empty-function
+const AsyncFunction = async function () {}.constructor as typeof Function
+
+export class ESModulesRunner implements ViteModuleRunner {
+  async runViteModule(
+    context: ViteRuntimeModuleContext,
+    transformed: string,
+  ): Promise<any> {
+    // use AsyncFunction instead of vm module to support broader array of environments out of the box
+    const initModule = new AsyncFunction(
+      ssrModuleExportsKey,
+      ssrImportMetaKey,
+      ssrImportKey,
+      ssrDynamicImportKey,
+      ssrExportAllKey,
+      // source map should already be inlined by Vite
+      '"use strict";' + transformed,
+    )
+
+    await initModule(
+      context[ssrModuleExportsKey],
+      context[ssrImportMetaKey],
+      context[ssrImportKey],
+      context[ssrDynamicImportKey],
+      context[ssrExportAllKey],
+    )
+
+    Object.freeze(context[ssrModuleExportsKey])
+  }
+
+  runExternalModule(filepath: string): Promise<any> {
+    return import(filepath)
+  }
+
+  processImport(
```

Comment:
I am on the fence about having this be part of the runner API - should we just always wrap externalized dependencies in this?

Reply:
I can't think of any other use for this method, so I think always wrapping is fine.